python/az/aro/azext_aro/custom.py

```diff
@@ -153,23 +163,40 @@ def aro_delete(cmd, client, resource_group_name, resource_name, no_wait=False):
             type='virtualNetworks',
             name=master_parts['name'],
         )
+        
+        route_tables = get_route_tables_from_subnets(cmd.cli_ctx, worker_subnets | {master_subnet})
 
         aad = AADManager(cmd.cli_ctx)
         rp_client_id = os.environ.get('AZURE_FP_CLIENT_ID', FP_CLIENT_ID)
         rp_client_sp = aad.get_service_principal(rp_client_id)
 
+
         # Customers frequently remove the RP's permissions, then cannot
         # delete the cluster. Where possible, fix this before attempting
         # deletion.
         if rp_client_sp:
-            sp_id = rp_client_sp.object_id
-            assign_contributor_to_vnet(cmd.cli_ctx, vnet, sp_id)
-            assign_contributor_to_routetable(cmd.cli_ctx,
-                                             worker_subnets | {master_subnet},
-                                             sp_id)
-    except (CloudError, HttpOperationError) as e:
-        # Default to old deletion behaviour in case operations throw an
-        # exception above. Log the error.
+            # Create role assignments if they don't exist
+            # Note: This logic assumes that if we cannot if the role assignment exists,
+            # we assume that it does.
+            vnet_contributor_exists = True  
+            try:
+                vnet_contributor_exists = has_contributor_on_vnet(cmd.cli_ctx, vnet, rp_client_sp.object_id)
+            except CloudError:
+                pass
+
+            if not vnet_contributor_exists:
+                assign_contributor_to_vnet(cmd.cli_ctx, vnet, rp_client_sp.object_id)
+            
+            for rt in route_tables:
+                route_table_contributor_exists = True
+                try:
+                    route_table_contributor_exists = has_contributor_on_routetable(cmd.cli_ctx, rt, rp_client_sp.object_id)
+                except CloudError:
+                    pass
+                
+                if not route_table_contributor_exists:
+                    assign_contributor_to_routetable(cmd.cli_ctx, rt, rp_client_sp.object_id)
+    except HttpOperationError as e:
```

Comment:
if `has_contributor_on_vnet` becomes `has_netowkr_contributor_on_object` 
and `assign_contributor_to_routetable` becomes `assign_network_contributor_to_object` (again, not sure if this is possible), this would become very similar to previous code with loop.

Reply:
Refactored and pushed.  Should be better now.